### PR TITLE
Batch delete source accounts

### LINKED ISSUE #1569: Support multiple select for sources
## Description
It would be helpful to be able to select multiple sources and then e.g. download all/delete all for the selected sources. At our organisation we frequently need to delete large numbers of messages so this would be a very helpful feature.

In terms of priority, this isn't urgent for us because we can use the old journalist portal to do this work at the moment.

As well as deletion and downloads, in future this feature could be used to support exporting the submissions/replies of multiple sources.

## How will this impact [SecureDrop users](https://github.com/freedomofpress/securedrop-ux/wiki/Users)?
As a securedrop workstation user, my organisation would find this very useful.

## How would this affect the SecureDrop Workstation [threat model](https://github.com/freedomofpress/securedrop-workstation/#threat-model)?
I don't think this changes the threat model as the behaviour is already possible by individually clicking on sources one by one.

## User Stories

As a user I want to be able to select multiple sources so that I can delete them in bulk.

### PR DESCRIPTION
**Pull request by zekehuntergreen**
# Description
Fixes #1569

- Add a checkbox to each source widget in the source list.
- Add a toolbar to the top of the source list with a button to delete all source accounts that are checked. The button is disabled by default and enabled whenever one or more sources is checked.
- Batch deletion of source accounts works similarly to deleting individual source accounts; clicking "Delete source accounts" opens a dialogue so that the user can confirm that they'd like to delete multiple source accounts. When they confirm, a DeleteSourceJob is added for each checked source widget. Users need to be signed in in order to delete multiple source accounts.

https://user-images.githubusercontent.com/17057932/217248635-96d579ea-2693-4491-906f-0783a16382f0.mov

# Test Plan

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [ ] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration is [self-contained] and applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [ ] No database schema changes are needed

[self-contained]: https://github.com/freedomofpress/securedrop-client#generating-and-running-database-migrations

### CODE REVIEW

File: securedrop_client/gui/widgets.py [hunk 1]
```diff
@@ -1303,20 +1342,26 @@ def __init__(
         self.paperclip_disabled.setSizePolicy(retain_space)
         self.paperclip_disabled.hide()
 
+        self.checkbox = QCheckBox()
+        self.checkbox.setObjectName("SourceWidget_checkbox")
+        self.checkbox.setStyleSheet(self.SOURCE_CHECKBOX_CSS)
+        self.checkbox.source_uuid = self.source.uuid
```
Comment (zekehuntergreen) [inline securedrop_client/gui/widgets.py]: might not be necessary

File: securedrop_client/logic.py [hunk 2]
```diff
@@ -418,6 +419,23 @@ def __init__(  # type: ignore [no-untyped-def]
...
+        Returns the list of sources that are checked in the UI.
+        """
+        source_items = self.gui.main_view.source_list.source_items
+        checked_source_uuids = []
+        for source_uuid, source_item in source_items.items():
+            source_item_widget = self.gui.main_view.source_list.itemWidget(source_item)
+            checkbox = source_item_widget.findChild(QCheckBox)
+            if checkbox.isChecked():
+                checked_source_uuids.append(source_uuid)
+        print("checked sources: ", checked_source_uuids)
```
Comment (zekehuntergreen) [inline securedrop_client/logic.py]: remove!

File: securedrop_client/gui/widgets.py [hunk 3]
```diff
@@ -642,8 +646,20 @@ def __init__(
         self.empty_conversation_view = EmptyConversationView()
         self.view_layout.addWidget(self.empty_conversation_view)
 
-        # Add widgets to layout
-        self._layout.addWidget(self.source_list, stretch=1)
+        self.sources_pane_holder = QWidget()
+        # TODO - remove this hard-coded width
```
Comment (philmcmahon) [inline securedrop_client/gui/widgets.py]: is this possible to remove it?

File: securedrop_client/logic.py [hunk 4]
```diff
@@ -1037,6 +1055,25 @@ def delete_source(self, source: db.Source) -> None:
...
+    def delete_sources(self, sources: List[str]) -> None:
+        """
+        Performs a delete operation on multiple source records.
+
+        This method will submit a job per source to delete the source record on
+        the server. If the job succeeds, the success handler will
+        synchronize the server records with the local state. If not,
+        the failure handler will display an error.
+        """
+        print(f"sources: {sources}")
```
Comment (philmcmahon) [inline securedrop_client/logic.py]: rogue log line

**Review by philmcmahon — APPROVED** [review]
Great work!

delete delete delete

File: securedrop_client/gui/widgets.py (hunk 3 above)
Comment (zekehuntergreen) [reply securedrop_client/gui/widgets.py]: I couldn't figure out another way :/
Without that style, on wide screens the toolbar is wider that the source list and there's an ugly grey space.

Screenshot 2023-02-08 at 11 50 30

Comment (philmcmahon) [reply securedrop_client/gui/widgets.py]: no problem

**Comment by gonzalo-bulnes** [comment]
@zekehuntergreen @philmcmahon We talked about it a while ago, if you're happy with this addition, I'd advocate for including it in FPF's release, and encourage you to open a PR against `freedomofpress/main` :slightly_smiling_face:

**Pull request merged**